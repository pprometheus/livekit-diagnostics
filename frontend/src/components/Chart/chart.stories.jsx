import React, { useState, useEffect } from "react";
import { DateTime } from "luxon";
import Chart from ".";

export default {
  title: "Components/Chart",
  component: Chart,
};

export const LiveChart = () => {
  const [data, setData] = useState(
    Array.from({ length: 8 }).map((_, i) => ({
      time: DateTime.now()
        .minus({ minutes: 7 - i })
        .toFormat("hh:mm a"),
      value: Math.random() * 100,
    }))
  );

  useEffect(() => {
    const interval = setInterval(() => {
      setData((old) => {
        const nextTime = DateTime.now().toFormat("hh:mm a");
        const nextValue = Math.random() * 100;
        return [...old.slice(1), { time: nextTime, value: nextValue }];
      });
    }, 1000); 

    return () => clearInterval(interval);
  }, []);

  return (
    <Chart
      title="Live Metrics"
      data={data}
      xDataKey="time"
      lines={[{ dataKey: "value", name: "Value", stroke: "#82ca9d" }]}
      yDomain={[0, 100]}
      unit="%"
      yTickFormatter={(value) => `${value}%`}
    />
  );
};
