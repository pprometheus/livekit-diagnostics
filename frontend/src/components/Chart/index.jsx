import { DateTime } from "luxon";
import React from "react";
import {
  ResponsiveContainer,
  LineChart,
  Line,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  Legend,
} from "recharts";

const Chart = ({
  title,
  data,
  xDataKey,
  lines,
  className = "",
  yDomain,
  unit,
  yTickFormatter,
}) => (
  <div
    className={`bg-gray-900 rounded-lg shadow p-4 flex flex-col flex-1 ${className}`}
  >
    <h3 className="text-gray-200 mb-2 text-sm font-semibold">{title}</h3>
    <div className="flex h-[200px]">
      <ResponsiveContainer width="100%" height="100%">
        <LineChart
          data={data}
          margin={{ top: 10, right: 20, left: 0, bottom: 5 }}
        >
          <CartesianGrid
            stroke="#2D3748"
            horizontal={true}
            vertical={false}
            strokeDasharray="0"
          />
          <XAxis
            dataKey={xDataKey}
            stroke="#718096"
            tick={{ fontSize: 12 }}
            tickLine={{ stroke: "transparent" }}
            axisLine={{ stroke: "#2D3748" }}
            padding={{ left: 20, right: 20 }}
            interval="preserveStartEnd"
            tickFormatter={(value) => {
              // Parse with Luxon and maintain HH:mm format
              const dt = DateTime.fromFormat(value, "hh:mm a");
              return dt.isValid ? dt.toFormat("hh:mm a") : value;
            }}
          />
          <YAxis
            orientation="right" // Move Y-axis to right side
            stroke="#718096"
            domain={yDomain}
            tick={{ fontSize: 12 }}
            tickLine={{ stroke: "transparent" }}
            axisLine={{ stroke: "#2D3748" }}
            tickFormatter={
              yTickFormatter || ((value) => `${value}${unit || ""}`)
            }
            width={40}
          />
          {/* Restored Tooltip */}
          <Tooltip
            contentStyle={{
              backgroundColor: "#1A202C",
              border: "none",
              borderRadius: "6px",
              boxShadow: "0 2px 4px rgba(0,0,0,0.2)",
            }}
            labelStyle={{
              color: "#CBD5E0",
              fontSize: 12,
              textAlign: "left",
            }}
            itemStyle={{
              color: "#fff",
              fontSize: 12,
            }}
            position={{ x: 0, y: 0 }} // Adjust tooltip position
          />
          {lines.map((line) => (
            <Line
              key={line.dataKey}
              type="monotone"
              dataKey={line.dataKey}
              name={line.name}
              stroke={line.stroke}
              strokeWidth={2}
              dot={false}
              activeDot={{ r: 4 }}
            />
          ))}
        </LineChart>
      </ResponsiveContainer>
    </div>
  </div>
);

export default Chart;
