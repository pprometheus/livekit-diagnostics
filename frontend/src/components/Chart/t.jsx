import React from "react";
import { Line } from "react-chartjs-2";
import {
  Chart as ChartJS,
  LineElement,
  PointElement,
  LinearScale,
  CategoryScale,
  TimeScale,
  Tooltip,
} from "chart.js";
import streamingPlugin from "chartjs-plugin-streaming"; // Import the streaming plugin
import "chartjs-adapter-moment"; // Import the moment adapter for Chart.js
import moment from "moment";

// Register required components and the streaming plugin
ChartJS.register(
  LineElement,
  PointElement,
  LinearScale,
  CategoryScale,
  TimeScale,
  Tooltip,
  streamingPlugin // Register the streaming plugin
);

const chartColors = {
  red: "rgb(255, 99, 132)",
  orange: "rgb(255, 159, 64)",
  yellow: "rgb(255, 205, 86)",
  green: "rgb(75, 192, 192)",
  blue: "rgb(54, 162, 235)",
  purple: "rgb(153, 102, 255)",
  grey: "rgb(201, 203, 207)",
};

const data = {
  datasets: [
    {
      label: "Dataset 1 (linear interpolation)",
      backgroundColor: "rgba(255, 99, 132, 0.5)", // Use RGBA directly
      borderColor: chartColors.red,
      fill: false,
      lineTension: 0,
      borderDash: [8, 4],
      data: [],
    },
  ],
};

const options = {
  elements: {
    line: {
      tension: 0.5,
    },
  },
  scales: {
    x: {
      type: "realtime", // Use the realtime scale provided by the streaming plugin
      realtime: {
        onRefresh: function (chart) {
          chart.data.datasets[0].data.push({
            x: moment(),
            y: Math.random(),
          });
        },
        delay: 3000,
        time: {
          displayFormat: "h:mm",
        },
      },
      ticks: {
        maxRotation: 0,
        minRotation: 0,
        stepSize: 1,
        maxTicksLimit: 30,
        source: "auto",
        autoSkip: true,
        callback: function (value) {
          return moment(value, "HH:mm:ss").format("mm:ss");
        },
      },
    },
    y: {
      ticks: {
        beginAtZero: true,
        max: 1,
      },
    },
  },
};

const Pp = () => {
  return (
    <div className="App">
      <Line data={data} options={options} />
    </div>
  );
};

export default Pp;
