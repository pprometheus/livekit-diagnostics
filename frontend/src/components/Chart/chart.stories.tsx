import React from "react";
import { DateTime } from "luxon";
import Chart from ".";

const meta = {
  title: "Components/Chart",
  component: Chart,
  tags: ["autodocs"],
};

export default meta;

const sampleData = [
  { time: "01:00 AM", value1: 10, value2: 15 },
  { time: "02:00 AM", value1: 20, value2: 25 },
  { time: "03:00 AM", value1: 15, value2: 30 },
  { time: "04:00 AM", value1: 25, value2: 35 },
];

export const Default = {
  args: {
    title: "Sample Line Chart",
    data: sampleData,
    xDataKey: "time",
    lines: [
      { dataKey: "value1", name: "Metric A", stroke: "#8884d8" },
      { dataKey: "value2", name: "Metric B", stroke: "#82ca9d" },
    ],
    unit: "",
  },
};

export const WithCustomYTicks = {
  args: {
    ...Default.args,
    title: "With Custom Y Tick Formatter",
    yTickFormatter: (value) => `$${value.toFixed(2)}`,
  },
};

export const SingleLine = {
  args: {
    title: "Single Line Chart",
    data: sampleData,
    xDataKey: "time",
    lines: [{ dataKey: "value1", name: "Only Metric", stroke: "#ff7300" }],
    unit: "",
  },
};
