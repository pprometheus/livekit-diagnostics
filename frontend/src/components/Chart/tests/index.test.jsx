import React from "react";
import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { DateTime } from "luxon";
import { Line, XAxis, YAxis, Tooltip } from "recharts";
import Chart from "..";

jest.mock("recharts", () => ({
  ResponsiveContainer: ({ children }) => children,
  LineChart: ({ children }) => <div data-testid="line-chart">{children}</div>,
  Line: jest.fn(() => <div>Line</div>),
  XAxis: jest.fn(() => <div>XAxis</div>),
  YAxis: jest.fn(() => <div>YAxis</div>),
  CartesianGrid: jest.fn(() => <div>CartesianGrid</div>),
  Tooltip: jest.fn(() => <div>Tooltip</div>),
  Legend: jest.fn(() => <div>Legend</div>),
}));

describe("Chart Component", () => {
  const mockData = [{ time: "10:00 AM", value: 100 }];
  const mockLines = [{ dataKey: "value", name: "Value", stroke: "#ff0000" }];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("renders title correctly", () => {
    render(<Chart title="Test Chart" data={[]} xDataKey="time" lines={[]} />);
    expect(screen.getByText("Test Chart")).toBeInTheDocument();
  });

  test("renders correct number of lines", () => {
    render(
      <Chart title="Test" data={mockData} xDataKey="time" lines={mockLines} />
    );
    expect(Line).toHaveBeenCalledTimes(mockLines.length);
    expect(Line.mock.calls[0][0].dataKey).toBe(mockLines[0].dataKey);
  });

  test("formats XAxis ticks correctly", () => {
    const xDataKey = "time";
    render(
      <Chart
        title="Test"
        data={mockData}
        xDataKey={xDataKey}
        lines={mockLines}
      />
    );

    const tickFormatter = XAxis.mock.calls[0][0].tickFormatter;
    const formattedTime = tickFormatter("10:00 AM");
    const expectedTime = DateTime.fromFormat("10:00 AM", "hh:mm a").toFormat(
      "hh:mm a"
    );
    expect(formattedTime).toBe(expectedTime);
  });

  test("configures YAxis domain and tickFormatter", () => {
    const yDomain = [0, 100];
    const yTickFormatter = (value) => `${value}%`;
    render(
      <Chart
        title="Test"
        data={mockData}
        xDataKey="time"
        lines={mockLines}
        yDomain={yDomain}
        yTickFormatter={yTickFormatter}
      />
    );

    const yAxisProps = YAxis.mock.calls[0][0];
    expect(yAxisProps.domain).toEqual(yDomain);
    expect(yAxisProps.tickFormatter(50)).toBe("50%");
  });

  test("applies correct styles to Tooltip", () => {
    render(
      <Chart title="Test" data={mockData} xDataKey="time" lines={mockLines} />
    );

    const tooltipProps = Tooltip.mock.calls[0][0];
    expect(tooltipProps.contentStyle.backgroundColor).toBe("#1A202C");
    expect(tooltipProps.labelStyle.color).toBe("#CBD5E0");
  });

  test("matches snapshot", () => {
    render(
      <Chart title="Test" data={mockData} xDataKey="time" lines={mockLines} />
    );
    const chartElement = screen.getByTestId("line-chart");
    expect(chartElement).toMatchSnapshot();
  });
});
