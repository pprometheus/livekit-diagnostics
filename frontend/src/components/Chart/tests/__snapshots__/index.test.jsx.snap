// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart Component matches snapshot 1`] = `
<div
  data-testid="line-chart"
>
  <div>
    CartesianGrid
  </div>
  <div>
    XAxis
  </div>
  <div>
    YAxis
  </div>
  <div>
    Tooltip
  </div>
  <div>
    Line
  </div>
</div>
`;
